### PR TITLE
Configurable pin

### PR DESCRIPTION
**Pull request by kewilli**
Working PR for yet another flavor of pin which checks the global state for existence, but does not spend the time to ensure the content is moved locally before returning. Expected semantics for a lightweight pin (I'm not attached the name):

* Lookup global existence for requested hashes
* Fire-and-forget a normal L2 pin, but do not await its result
* Return the result from the global existence check such that content found _anywhere_ returns `Success`

### CODE REVIEW

File: Public/Src/Cache/ContentStore/Distributed/Sessions/ILightweightPin.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+{
+    /// <summary>
+    /// Pin in a (hopefully) faster way, while being less reliable.
+    /// </summary>
+    public interface ILightweightPin
+    {
+        /// <summary>
+        /// Check for existence in L2, return the missing content, then fire-and-forget L2 pin to locally copy the found content.
+        /// </summary>
+        Task<IEnumerable<Task<Indexed<PinResult>>>> LightweightPinAsync(OperationContext operationContext, IReadOnlyList<ContentHash> contentHashes, UrgencyHint urgencyHint);
```
Comment (kewilli) [inline Public/Src/Cache/ContentStore/Distributed/Sessions/ILightweightPin.cs]: PinAsync(OperationContext, hashes, PinOperationConfiguration configureation);

PinOperationConfiguration {
   public bool 
}

File: Public/Src/Cache/ContentStore/Distributed/Sessions/ReadOnlyDistributedContentSession.cs [hunk 2]
```diff
@@ -360,17 +360,53 @@ async Task<BoolResult> TryCopyContentLocalAsync()
...
-            return await Workflows.RunWithFallback(
-                contentHashes,
-                hashes => Inner.PinAsync(operationContext, hashes, operationContext.Token, urgencyHint),
-                hashes => _remotePinner(operationContext, hashes, operationContext.Token, urgencyHint),
-                result => result.Succeeded,
-                // Exclude the empty hash because it is a special case which is hard coded for place/openstream/pin.
-                async hits => await UpdateContentTrackerWithLocalHitsAsync(operationContext, hits.Where(x => !(Settings.EmptyFileHashShortcutEnabled && contentHashes[x.Index].IsEmptyHash())).Select(x => new ContentHashWithSizeAndLastAccessTime(contentHashes[x.Index], x.Item.ContentSize, x.Item.LastAccessTime)).ToList(), operationContext.Token, urgencyHint));
+            IEnumerable<Task<Indexed<PinResult>>> pinResults = null;
+
+            if (pinOperationConfiguration.ReturnGlobalExistenceFast)
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Distributed/Sessions/ReadOnlyDistributedContentSession.cs]: I think this should be done in the second callback provided to RunWithFallback.

Currently, you go to the remote first without trying to pin locally.

File: Public/Src/Cache/ContentStore/Distributed/Sessions/ReadOnlyDistributedContentSession.cs [hunk 3]
```diff
@@ -360,17 +360,59 @@ async Task<BoolResult> TryCopyContentLocalAsync()
...
+            {
+                // Check globally for existence, but do not copy locally
+                pinResults = await Workflows.RunWithFallback(
+                    contentHashes,
+                    hashes => Inner.PinAsync(operationContext, hashes, operationContext.Token, urgencyHint),
+                    hashes => _remotePinner(operationContext, hashes, operationContext.Token, true, urgencyHint),
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Distributed/Sessions/ReadOnlyDistributedContentSession.cs]: nit: could you please use named arguments for 'true' and 'false'?

File: Public/Src/Cache/ContentStore/Distributed/Sessions/ReadOnlyDistributedContentSession.cs [hunk 4]
```diff
@@ -360,17 +360,59 @@ async Task<BoolResult> TryCopyContentLocalAsync()
...
+            return PinHelperAsync(operationContext, contentHashes, UrgencyHint.Nominal, pinOperationConfiguration);
+        }
+
+        /// <inheritdoc />
+        protected override Task<IEnumerable<Task<Indexed<PinResult>>>> PinCoreAsync(OperationContext operationContext, IReadOnlyList<ContentHash> contentHashes, UrgencyHint urgencyHint, Counter retryCounter, Counter fileCounter)
+        {
+            return PinHelperAsync(operationContext, contentHashes, urgencyHint, PinOperationConfiguration.Default());
+        }
+
+        private async Task<IEnumerable<Task<Indexed<PinResult>>>> PinHelperAsync(OperationContext operationContext, IReadOnlyList<ContentHash> contentHashes, UrgencyHint urgencyHint, PinOperationConfiguration pinOperationConfiguration)
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Distributed/Sessions/ReadOnlyDistributedContentSession.cs]: nit: but I would suggest to move urgency hint into the pin configuration.

File: Public/Src/Cache/ContentStore/Distributed/Stores/DistributedContentCopier.cs [hunk 5]
```diff
@@ -348,12 +348,12 @@ public Task<BoolResult> RequestCopyFileAsync(OperationContext context, ContentHa
...
                     {
                         // Handles both OperationCanceledException and TaskCanceledException (TaskCanceledException derives from OperationCanceledException)
-                        return reportCancellationRequested();
+                        return reportCancellationRequested("Exception: " + e);
                     }
 
                     if (cts.IsCancellationRequested)
                     {
-                        return reportCancellationRequested();
+                        return reportCancellationRequested("Token cancelled 2");
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Distributed/Stores/DistributedContentCopier.cs]: Are you planning to keep this?

File: Public/Src/Cache/DistributedCache.Host/Service/Internal/MultiplexedContentSession.cs [hunk 6]
```diff
@@ -24,6 +27,17 @@ public MultiplexedContentSession(ContentSessionTracer tracer, Dictionary<string,
         {
         }
 
+        public Task<IEnumerable<Task<Indexed<PinResult>>>> LightweightPinAsync(OperationContext operationContext, IReadOnlyList<ContentHash> contentHashes, UrgencyHint urgencyHint)
+        {
+            var session = PreferredContentSession as ILightweightPin;
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/DistributedCache.Host/Service/Internal/MultiplexedContentSession.cs]: It seems that this part is not updated yet.

File: Public/Src/Cache/ContentStore/Distributed/Sessions/ReadOnlyDistributedContentSession.cs [hunk 7]
```diff
@@ -696,25 +738,37 @@ private Task<BoolResult> UpdateContentTrackerWithNewReplicaAsync(Context context
             Context context,
             IReadOnlyList<ContentHash> contentHashes,
             CancellationToken cts,
+            bool onlyCheckGlobalExistence,
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Distributed/Sessions/ReadOnlyDistributedContentSession.cs]: >onlyCheckGlobalExistence [](start = 17, length = 24)

could you please use named args for boolean literals here.

File: Public/Src/Cache/ContentStore/Distributed/Sessions/ReadOnlyDistributedContentSession.cs [hunk 8]
```diff
@@ -696,25 +738,37 @@ private Task<BoolResult> UpdateContentTrackerWithNewReplicaAsync(Context context
...
-            return Workflows.RunWithFallback(
-                contentHashes,
-                hashes => PinFromContentLocationStoreOriginAsync(operationContext, hashes, cts, GetBulkOrigin.Local, urgencyHint),
-                hashes => PinFromContentLocationStoreOriginAsync(operationContext, hashes, cts, GetBulkOrigin.Global, urgencyHint),
-                result => result.Succeeded);
+            if (onlyCheckGlobalExistence)
+            {
+                return Workflows.RunWithFallback(
+                    contentHashes,
+                    hashes => PinFromContentLocationStoreOriginAsync(operationContext, hashes, cts, GetBulkOrigin.Global, true, urgencyHint),
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Distributed/Sessions/ReadOnlyDistributedContentSession.cs]: In this case you always will be hitting the redis.

You still can look into the local location store here as well...

Lets discuss this.

File: Public/Src/Cache/DistributedCache.Host/Service/Internal/MultiplexedContentSession.cs [hunk 9]
```diff
@@ -24,6 +27,17 @@ public MultiplexedContentSession(ContentSessionTracer tracer, Dictionary<string,
         {
         }
 
+        public Task<IEnumerable<Task<Indexed<PinResult>>>> PinAsync(OperationContext operationContext, IReadOnlyList<ContentHash> contentHashes, PinOperationConfiguration pinOperationConfiguration)
+        {
+            var session = PreferredContentSession as IConfigurablePin;
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/DistributedCache.Host/Service/Internal/MultiplexedContentSession.cs]: BTW, I think this local variable is not needed.

**Pull request merged**